File: src/judge/python3/judge.py
import json
from subprocess import Popen, PIPE, TimeoutExpired
import io
from time import time, sleep

#Create subprocess here, so subprocess can run init code first. 
def judge_handler(event, context):
    print(event)
    print('=================Run subprocess=================')
    proc = Popen(['python', 'run.py'], stdin=PIPE, stdout=PIPE)
    sleep(0.09) #When run in 2048mb lambda, 0.07 is enough to finish init code.
    print('=================Subprocess Ready===============')

    data = json.dumps(event).encode('utf8')
    t = -time()
    try:
        outs, errs = proc.communicate(
            input=data,
            timeout=int(context.get_remaining_time_in_millis())
        )
        t += time()
    except TimeoutExpired:
        proc.kill()
        outs, errs = proc.communicate()
    outs = outs.decode('utf8')
    result = json.loads(outs)
    if t > 0:
        result['run_duration'] = t * 1000
    print('=================Subprocess end=====')
    print('OUT>>>>>>>>\n{}\nERROR>>>>>>>>\n{}'.format(result, errs))

    proc = Popen(['python', 'run.py'], stdin=PIPE, stdout=PIPE)
    return result
    

File: src/judge/nodejs/judge.js
const cp = require('child_process')

exports.judge_handler = function(event, context, callback) {
    console.log('=================Run subprocess=================')
    const n = cp.fork('./subprocess.js')
    
    function ret(result){
        n.kill('SIGINT')
        callback(null, result)
    } 
    
    let timeout = setTimeout(()=>{
        ret({
            error: 'TLE'
        })
    }, event.timeout)
    let t = 0
    n.on('message', (m) => {
        if(m == 'READY' && t == 0){
            console.log('=================Subprocess Ready===============');
            t = - process.uptime()
            n.send(event)
        }else{
            t += process.uptime()
            clearTimeout(timeout)
            console.log('=================Get Result=====================');
            console.log(m)
            m.run_duration = Math.round(t * 1000)
            ret(m)
            return
        }
    });
}

///////////////////////////////////////////
//event = {
//    params:[ 
//        { name: 'a', type: 'int', description: 'A' },
//        { name: 'b', type: 'int', description: 'B' }
//    ],
//    timeout: 3000,
//    method: 'aplusb',
//    testcase: [ [ 1, 2 ], [ 2, 2 ] ],
//    code: 'function aplusb(a, b){\n console.log(`${a} + ${b} = ${a+b}`)\n console.error(`${a} + ${b} = ${a+b}`)\n throw \'Guck\'\n return a + b + 1\n}'
//}
//
//
//context = {}
//callback = (err, result) =>{
//    console.log('============@@@@@@@=================')
//    console.log('err', err)
//    console.log('result', result)
//}
//
//exports.judge_handler(event, context, callback)